### PR TITLE
feat: add middleware support

### PR DESCRIPTION
**Pull request by qingzhuozhen**
### Summary

- add middleware support
- add unit tests

### Checklist

* [ ] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-iOS/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:

### CODE REVIEW

File: Sources/Amplitude/Amplitude.m [hunk 1]
```diff
@@ -513,6 +517,10 @@ - (void)logEvent:(NSString *)eventType withEventProperties:(NSDictionary *)event
     [self logEvent:eventType withEventProperties:eventProperties withGroups:nil];
 }
 
+- (void)logEvent:(NSString *)eventType withEventProperties:(nullable NSDictionary *)eventProperties extra: (nullable NSMutableDictionary *) extra {
+    [self logEvent:eventType withEventProperties:eventProperties withApiProperties:nil withUserProperties:nil withGroups:nil withGroupProperties:nil withTimestamp:nil outOfSession:NO extra:extra];
```
Comment (justin-fiedler) [inline Sources/Amplitude/Amplitude.m:521]: Besides `outOfSession` it seems like all params start with "with". Could be good to stay consistent e.g. `withExtra` or `withMiddlewareExtra`

File: Tests/AmplitudeTests.m [hunk 2]
```diff
@@ -1110,17 +1110,45 @@ - (void)testSetPlan {
...
+    [client initializeApiKey:@"eu_api_key"];
     [client setServerZone:EU];
     XCTAssertEqualObjects(kAMPEventLogEuUrl, [client valueForKey:@"serverUrl"]);
 }
 
 - (void)testSetServerZoneWithoutUpdateServerUrl {
     Amplitude *client = [Amplitude instanceWithName:@"eu_zone_2"];
     XCTAssertEqualObjects(kAMPEventLogUrl, [client valueForKey:@"serverUrl"]);
-    [client initializeApiKey:@"eu_api_ket"];
+    [client initializeApiKey:@"eu_api_key"];
```
Comment (justin-fiedler) [inline Tests/AmplitudeTests.m:1121]: 👍

File: Sources/Amplitude/AMPMiddleware.m [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+@implementation AMPBlockMiddleware
+
+- (instancetype _Nonnull)initWithBlock:(AMPMiddlewareBlock)block {
+    if (self = [super init]) {
+        _block = block;
+    }
+    return self;
+}
+
+- (void)proceed:(AMPMiddlewarePayload *)payload next:(AMPMiddlewareNext)next {
```
Comment (justin-fiedler) [inline Sources/Amplitude/AMPMiddleware.m:48]: I think we could rename `proceed` => `run` to align with the MiddlewareRunner.

**Review by justin-fiedler — CHANGES_REQUESTED** [review]

File: Sources/Amplitude/Amplitude.m (hunk 1 above)
Comment (qingzhuozhen) [reply Sources/Amplitude/Amplitude.m]: Renamed the param to `withMiddlewareExtra`.

File: Sources/Amplitude/AMPMiddleware.m (hunk 3 above)
Comment (qingzhuozhen) [reply Sources/Amplitude/AMPMiddleware.m]: renamed. At first I was trying to make it a bit different to separate. I guess it makes sense to align them.

File: Tests/MiddlewareRunnerTests.m [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    }];
+    [_middlewareRunner add:updateEventTypeMiddleware];
+    [_middlewareRunner add:updateDeviceModelMiddleware];
+    
+    NSMutableDictionary *event = [NSMutableDictionary dictionary];
+    [event setValue:@"sample_event" forKey:@"event_type"];
+    [event setValue:@"sample_device" forKey:@"device_model"];
+    NSMutableDictionary *extra = [NSMutableDictionary dictionary];
+    
+    AMPMiddlewarePayload * middlewarePayload = [[AMPMiddlewarePayload alloc] initWithEvent:event extra:extra];
```
Comment (justin-fiedler) [inline Tests/MiddlewareRunnerTests.m:49]: Nit: withExtra?

**Review by justin-fiedler — CHANGES_REQUESTED** [review]

File: Tests/MiddlewareRunnerTests.m (hunk 4 above)
Comment (qingzhuozhen) [reply Tests/MiddlewareRunnerTests.m]: yeah, just updated for MiddlewarePayload as well.

**Review by justin-fiedler — APPROVED** [review]

File: Tests/MiddlewareRunnerTests.m [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    [_middlewareRunner run:middlewarePayload next:^(AMPMiddlewarePayload *_Nullable newPayload){
+        middlewareCompleted = YES;
+    }];
+    
+    XCTAssertEqual(middlewareCompleted, YES);
+    XCTAssertEqualObjects([event objectForKey:@"event_type"], eventType);
+    XCTAssertEqualObjects([event objectForKey:@"device_model"], deviceModel);
+}
+
+- (void)testRunWithNotPassMiddleware {
```
Comment (yuhao900914) [inline Tests/MiddlewareRunnerTests.m:62]: testRunWithoutPassMiddleware?

Comment (qingzhuozhen) [reply Tests/MiddlewareRunnerTests.m:62]: `NotPassMiddleware` is the name of middleware, trying to test if it stops on further action

**Review by yuhao900914 — APPROVED** [review]

**Pull request merged**